In [1]:
from pexl.io.converter import xlsx_to_dataset_dir

version = "v1.10_4"

xlsx_to_dataset_dir(
    f"../data/exports/schema_{version}.xlsx", # created by PEExcel macro "export_schema"
    f"../data/schemas/{version}", #creates this directory and csv files
    mode="schema",
    replace_files=True # careful!
)



ConvertResult(out_dir=WindowsPath('../data/schemas/v1.10_4'), written={'IN': WindowsPath('../data/schemas/v1.10_4/IN.csv'), 'OUT': WindowsPath('../data/schemas/v1.10_4/OUT.csv'), 'SIM': WindowsPath('../data/schemas/v1.10_4/SIM.csv')}, audit_path=WindowsPath('../data/schemas/v1.10_4/AUDIT.json'))

In [2]:
from pexl.validate.schema import diff_schema_dirs

diff_result = diff_schema_dirs(
    old_dir=f"../data/schemas/v1.09",
    new_dir= f"../data/schemas/{version}"
)
diff_result

{'meta': {'old_dir': '..\\data\\schemas\\v1.09',
  'new_dir': '..\\data\\schemas\\v1.10_4',
  'key_col': 'var_name'},
 'tables': {'IN': {'columns': {'added': [], 'removed': [], 'common_count': 8},
   'rows': {'old_count': 579,
    'new_count': 612,
    'new_errors': 6,
    'var_name_added': ['COMP_area_balconies',
     'COMP_area_basement_ceiling',
     'COMP_area_basement_floor_unheated',
     'COMP_area_ceil_to_air',
     'COMP_area_ceiling_topfloor',
     'COMP_area_columns',
     'COMP_area_fundament',
     'COMP_area_internal_wall_load',
     'COMP_area_internal_wall_nonload',
     'COMP_area_internal_wall_unheated',
     'COMP_area_terrace',
     'COMP_area_unheated_horizontal',
     'COMP_area_wall_earth_contacted',
     'COMP_area_wall_ec_unheated',
     'COMP_area_windowframe',
     'COMP_area_windowglazing',
     'COMP_name_balconies',
     'COMP_name_basement_ceiling',
     'COMP_name_basement_floor_unheated',
     'COMP_name_ceil_to_air',
     'COMP_name_ceiling_topfloor',


In [3]:
from pexl.validate.schema import schema_diff_to_markdown


md = schema_diff_to_markdown(diff_result, value_format="None")
print(md)

# Schema Diff

- **Old:** `..\data\schemas\v1.09`
- **New:** `..\data\schemas\v1.10_4`
- **Key:** `var_name`

## IN

### Columns
- Added: []
- Removed: []
- Common count: 8

### Rows
Old count: 579
New count: 612
- **New Error count: 6**
- Added/~~Removed~~ var_name:
   - COMP_area_balconies
  - COMP_area_basement_ceiling
  - COMP_area_basement_floor_unheated
  - COMP_area_ceil_to_air
  - COMP_area_ceiling_topfloor
  - COMP_area_columns
  - COMP_area_fundament
  - COMP_area_internal_wall_load
  - COMP_area_internal_wall_nonload
  - COMP_area_internal_wall_unheated
  - COMP_area_terrace
  - COMP_area_unheated_horizontal
  - COMP_area_wall_earth_contacted
  - COMP_area_wall_ec_unheated
  - COMP_area_windowframe
  - COMP_area_windowglazing
  - COMP_name_balconies
  - COMP_name_basement_ceiling
  - COMP_name_basement_floor_unheated
  - COMP_name_ceil_to_air
  - COMP_name_ceiling_topfloor
  - COMP_name_columns
  - COMP_name_fundament
  - COMP_name_internal_wall_load
  - COMP_name_internal_w

# Schema Diff

- **Old:** `..\data\schemas\v1.09`
- **New:** `..\data\schemas\v1.10_4`
- **Key:** `var_name`

## IN

### Columns
- Added: []
- Removed: []
- Common count: 8

### Rows
Old count: 579
New count: 612
- **New Error count: 6**
- Added/~~Removed~~ var_name:
   - COMP_area_balconies
  - COMP_area_basement_ceiling
  - COMP_area_basement_floor_unheated
  - COMP_area_ceil_to_air
  - COMP_area_ceiling_topfloor
  - COMP_area_columns
  - COMP_area_fundament
  - COMP_area_internal_wall_load
  - COMP_area_internal_wall_nonload
  - COMP_area_internal_wall_unheated
  - COMP_area_terrace
  - COMP_area_unheated_horizontal
  - COMP_area_wall_earth_contacted
  - COMP_area_wall_ec_unheated
  - COMP_area_windowframe
  - COMP_area_windowglazing
  - COMP_name_balconies
  - COMP_name_basement_ceiling
  - COMP_name_basement_floor_unheated
  - COMP_name_ceil_to_air
  - COMP_name_ceiling_topfloor
  - COMP_name_columns
  - COMP_name_fundament
  - COMP_name_internal_wall_load
  - COMP_name_internal_wall_nonload
  - COMP_name_internal_wall_unheated
  - COMP_name_other
  - COMP_name_terrace
  - COMP_name_unheated_horizontal
  - COMP_name_wall_earth_contacted
  - COMP_name_wall_ec_unheated
  - COMP_name_windowframe
  - COMP_name_windowglazing
  - EV_mileage_residential
  - GPW_ceilings_name
  - Plugloads_scale_residential
  - QHmax_1el_m2
  - QHmax_2th_m2
  - QHmax_3el_m2
  - QHmax_4th_m2
  - StatPAX_residential
  - StatPAX_schoolprim
  - StatPAX_schoolsec
  - density_NFApPers_schoolprim
  - density_NFApPers_schoolsec
  - fc_schoolprim
  - fc_schoolsec
  - input_version_date
  - transmittance_Wm2
  - usage_concurrency_summer_residential
  - usage_concurrency_summer_schoolprim
  - usage_concurrency_summer_schoolsec
  - usage_concurrency_winter_residential
  - usage_concurrency_winter_schoolprim
  - usage_concurrency_winter_schoolsec
  - ~~EV_mileage_res~~
  - ~~GWP_ceilings_name~~
  - ~~Plugloads_scale_res~~
  - ~~QHmax_1el_m²~~
  - ~~QHmax_2th_m²~~
  - ~~QHmax_3el_m²~~
  - ~~QHmax_4th_m²~~
  - ~~StatPAX_eduprim~~
  - ~~StatPAX_edusec~~
  - ~~StatPAX_res~~
  - ~~density_NFApPers_edu~~
  - ~~density_NFApPers_edu_rpim~~
  - ~~fc_eduprim~~
  - ~~fc_edusec~~
  - ~~ghg_conversion_factor_gasoline~~
  - ~~transmittance_Wm²~~
  - ~~usage_concurrency_summer_eduprim~~
  - ~~usage_concurrency_summer_edusec~~
  - ~~usage_concurrency_summer_res~~
  - ~~usage_concurrency_winter_eduprim~~
  - ~~usage_concurrency_winter_edusec~~
  - ~~usage_concurrency_winter_res~~


### Potential Errors (#)

| var_name | column | error |
|---|---|---|
| fGHG_grid_column | Formel | VERGLEICH(fGHG_grid_profile;fGHG[#Kopfzeilen];0) |
| fPE_grid_column | Formel | VERGLEICH(fPE_grid_profile;fPE[#Kopfzeilen];0) |
| flex_Signals_selected_column | Formel | VERGLEICH(FLEX_signal_name;Signals[#Kopfzeilen];0) |
| rcp1_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp1_renewable;#NV);rcp1_fossil) |
| rcp2_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp2_renewable;#NV);rcp2_fossil) |
| rcp3_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp3_renewable;#NV);rcp3_fossil) |





## OUT

### Columns
- Added: []
- Removed: []
- Common count: 8

### Rows
Old count: 391
New count: 410
- **New Error count: 0**
- Added/~~Removed~~ var_name:
   - Batt_charging_losses
  - FSI
  - Ta_annual_avg
  - Ta_avg_cooling_period
  - Ta_avg_heating_period
  - Tc_avg_summer
  - Tc_avg_winter
  - Tu_avg_summer
  - Tu_avg_winter
  - dTc_summer
  - dTc_winter
  - dTu_summer
  - dTu_winter
  - lc
  - location_address
  - project_creation_date
  - project_name
  - project_scenario_name
  - renovation_share
  - rh_avg
  - test_heat_balance
  - weather_index
  - weather_name
  - ~~dTc summer~~
  - ~~dTc winter~~
  - ~~dTu summer~~
  - ~~dTu winter~~


### Potential Errors (#)
- None





## SIM

### Columns
- Added: []
- Removed: []
- Common count: 2

### Rows
Old count: 394
New count: 394
- **New Error count: 3**
- Added/~~Removed~~ var_name:
   - Qh_u2
  - Spalte2
  - Spalte4
  - Spalte8
  - ~~Batt_interhour_exchange~~
  - ~~Ecar_final_Whm2~~
  - ~~dT~~
  - ~~dT_validation~~


### Potential Errors (#)

| var_name | column | error |
|---|---|---|
| Signal | Formula | =INDEX(Signals,ROW()-ROW(Signals[#Headers]),flex_Signals_selected_column) |
| Ti0cooled | Formula | =IF(NFA_cooled/NFA_total>0,Tsetheat_min,#N/A) |
| Ti0uncooled | Formula | =IF(1-NFA_cooled/NFA_total>0,Tsetheat_min,#N/A) |





## Audit summary

- Raw audit present (old): True
- Raw audit present (new): True

### Empty var_name exclusions (converter)

**Old:**
```json
{'IN': {'empty_var_name_count': 109}, 'OUT': {'empty_var_name_count': 117}, 'SIM': {'empty_var_name_count': 0}}
```
**New:**
```json
{'IN': {'empty_var_name_count': 88}, 'OUT': {'empty_var_name_count': 98}, 'SIM': {'empty_var_name_count': 0}}
```

In [5]:

from pexl.io.schema_codegen import generate_schema_module


generate_schema_module(
    excel_path=f"../data/exports/schema_{version}.xlsx",
    output_py_path=f"../src/pexl/schemas/generated/excel_{version.replace(".", "_")}.py",
    version=version.replace(".", "_"), # so filename is valid python module (no dots)
)

import pexl.schemas.current as current_schema # latest schema must be manually set
current_schema.ExcelNamedVariables # is used by module excel 

pexl.schemas.generated.excel_v1_10_4.ExcelNamedVariables

In [9]:
from pexl.utils.excel import read_project


project = read_project("../data/exports/ka_project_backup_v1.10.xlsx")
project

<Project districts=27 warnings=59 source='..\\data\\exports\\ka_project_backup_v1.10.xlsx'>

In [10]:
project.districts

[<District 'Default' scenarios=1>,
 <District 'Gründerzeit' scenarios=3>,
 <District 'Volksschule Neubau' scenarios=1>,
 <District 'Wohnen Neubau 2.5 Geschosse' scenarios=1>,
 <District 'mg. Wohnbau 90er Sanierung' scenarios=1>,
 <District 'Mischnutzung Neubau 7-stöckig' scenarios=1>,
 <District 'AT_EFHo' scenarios=1>,
 <District 'Südtiroler Siedlung' scenarios=1>,
 <District 'EFHo' scenarios=5>,
 <District 'EFHk' scenarios=5>,
 <District 'MFHo' scenarios=4>,
 <District 'MFHk' scenarios=5>,
 <District 'GWBk' scenarios=5>,
 <District 'GWBg' scenarios=5>,
 <District 'MFHo |1945 - 1989 - existing - Erdgas' scenarios=1>,
 <District 'DLG1k' scenarios=9>,
 <District 'DLG1g' scenarios=9>,
 <District 'DLG2k' scenarios=9>,
 <District 'DLG2g' scenarios=9>,
 <District 'DLG3k' scenarios=9>,
 <District 'DLG3g' scenarios=9>,
 <District 'Südtiroler Quartier' scenarios=3>,
 <District 'RUGInnovation' scenarios=16>,
 <District 'Beispiel: Sanierung mg. Wohnbau' scenarios=2>,
 <District 'Am Bichl III' sce

In [16]:
d = project.get("EFHo")
print(d)
d.scenarios

<District 'EFHo' scenarios=5>


[<Scenario 'bis 1944 - existing - Erdgas'>,
 <Scenario '1945 - 1989 - existing - Erdgas'>,
 <Scenario '1990 - 1999 - existing - Erdgas'>,
 <Scenario '2000 - 2020 - existing - Erdgas'>,
 <Scenario 'nach 2020 - existing - Erdgas'>]

In [ ]:
a = d.scenarios[0]
b = d.get("nach 2020 - existing - Erdgas")
a,b

(<Scenario 'nach 2020 - existing - Erdgas'>,
 <Scenario 'bis 1944 - existing - Erdgas'>)